tests/meta/test_adaptive_random_forest_regressor.py

```diff
@@ -0,0 +1,52 @@
+import os
+import numpy as np
+from array import array
+from sklearn.metrics import mean_absolute_error
+from skmultiflow.data import RegressionGenerator
+from skmultiflow.meta import AdaptiveRandomForestRegressor
+
+
+def test_adaptive_random_forest_regressor():
+    stream = RegressionGenerator(n_samples=500, n_features=20, n_informative=15, random_state=1)
+    stream.prepare_for_use()
+
+    learner = AdaptiveRandomForestRegressor(random_state=1)
+
+    cnt = 0
+    max_samples = 500
+    y_pred = array('d')
+    y_true = array('d')
+    wait_samples = 10
+
+    while cnt < max_samples:
+        X, y = stream.next_sample()
+        # Test every n samples
+        if (cnt % wait_samples == 0) and (cnt != 0):
+            y_pred.append(learner.predict(X)[0])
+            y_true.append(y[0])
+        learner.partial_fit(X, y)
+        cnt += 1
+
+    error = mean_absolute_error(y_true, y_pred)
+    expected_error = 155.5787786499754
+    assert (np.abs(expected_error - error) < 1)
```

Comment:
I think that np.isclose would be better in this case.

Reply:
I had some trouble with that: the result seemed to vary quite a lot depending on the version of python & the OS - `the tests/utils/test_utils.py::test_calculate_object_size` test fail on my machines, too. Thus, I wasn't sure which tolerance to put in the `np.isclose` test. Looking for advice on that one!

Reply:
There is an issue for `test_calculate_object_size`. However, this should not affect the performance of the regressor. In the past, we had issues with the internal random number generators not being correctly set which caused errors during testing.

Reply:
I have committed a fix for the test 74c1bed3dc4c3af15c97e723ab57708de8ee4434 given that it has been identified to affect macOS primarily.

Reply:
![house16H_5000-datapoints](https://user-images.githubusercontent.com/19614225/70671264-b9ae8980-1c7b-11ea-9de0-4c27dc100af7.png)

This is what I got when training on 5000 data points of the house16H.arff dataset. The most notable difference is that error spikes, as if each trees error was the same & it was all added up. (I'll clean up the test & put it here asap.)
The MAE is comparable to that of a tree whilst no such error spike is encountered.
I'll compare with MOA's implem tomorrow to look for any functionnal difference that I missed.
I still have NaN errors, probably because of almost null sd that make NaN when normalising.

Also, test_calculate_object_size still fails here (Arch Linux).

Reply:
Thanks for the information. One thing to take into account is that the random number seed is properly being propagated. Otherwise, you will get different ensembles on each run and will be unable to test the object size calculation.

Reply:
![house16H_5000-datapoints_nestimators1](https://user-images.githubusercontent.com/19614225/70817684-57b46800-1dd2-11ea-8960-4ad9c181a0dc.png)
This is with n_estimators=1... Time to debug, I guess.

Besides, I pass `self.random_state` to the trees constructor, which seems correct.

Reply:
Interesting: when normalising, the variance is frequently close to 0. I haven't been able to reproduce this with the tree alone, so I assume it stems from the poisson weights of the forest. It creates some weird behaviour because we do not always check that case.

Reply:
Note: ADWIN also complains *a lot* about invalid values in sqrt.

Reply:
Hello to everyone! Some valid points to consider are the differences in the strategies for normalizing features and targets used in the tree algorithms.

FIMT-DD centers the observations by the average feature value and scales them by three times the standard deviation. iSOUP-Tree and its variants use the standard z-score approach: centering by the mean and scaling by the standard deviation. The argument for the latter approach, as stated in the original proposal (here I quote Osojnik's thesis), is that:

"Ikonomovska, Gama, and Džeroski (2011b) suggest a normalization in which the difference of the value and the mean is instead divided with 3σ. However, this has the exact effect as dividing the initial learning rate by 3. Therefore, modifying the normalization procedure produces the same results as modifying the learning rate. Thus, we use the standard normalization procedure and allow a user to modify the learning rate, which already appears as a parameter of the iSOUP-Tree model." 

*Structured Output Prediction on Data Streams*. Aljaž Osojnik. PhD Thesis. Jožef Stefan International Postgraduate School. Available [here](http://kt.ijs.si/theses/phd_aljaz_osojnik.pdf).

In the case where the standard deviation is smaller than one, multiplying it by three and using the result in the denominator of the normalization might result in big numbers.

Reply:
To have a baseline comparison, I ran my tests with a forest with 'mean' & no ADWIN. The results are as expected. So now, it's about 1/ 'perceptron' (surely the normalization correction will help tremendously in avoiding big numbers aberrations) 2/ ADWIN.

![house16H_5000-50-mean-no-ADWIN](https://user-images.githubusercontent.com/19614225/71081893-61eab380-2190-11ea-94dd-9204801608aa.png)

Reply:
Update: the 'perceptron' version works fine without ADWIN...